tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetUnionRule.scala

```diff
@@ -61,35 +60,25 @@ class SetUnionRule(rewriter: SymbStateRewriter) extends RewritingRule {
           //
           // This approach is more expensive at the rewriting phase, but it produces O(n) constraints in SMT,
           // in contrast to the old approach with equalities and uninterpreted functions, which required O(n^2) constraints.
-          def addOneElemCons(elemsCells: Seq[ArenaCell]): Unit = {
-            if (elemsCells.nonEmpty) {
-              def consChain(elems: Seq[ArenaCell]): BuilderEx =
-                ConsChainUtil.consChainFold[ArenaCell](
-                    elems,
-                    newSetCell.toNameEx,
-                    { elemCell =>
-                      def isPointedBySet(set: ArenaCell, setElems: Set[ArenaCell]): Boolean =
-                        setElems.contains(elemCell)
-                      def inPointingSet(set: ArenaCell) = {
-                        // this is sound, because we have generated element equalities
-                        // and thus can use congruence of in(...) for free
-                        tla.and(tla.apalacheSelectInSet(set.toNameEx, topSetCell.toNameEx),
-                            tla.apalacheSelectInSet(elemCell.toNameEx, set.toNameEx))
-                      }
-                      val pointingSets = (sets.zip(elemsOfSets) filter (isPointedBySet _).tupled) map (_._1)
-                      val inUnion = tla.apalacheStoreInSet(elemCell.toNameEx, newSetCell.toNameEx)
-                      val existsIncludingSet = tla.or(pointingSets map inPointingSet: _*)
-                      (inUnion, existsIncludingSet)
-                    },
-                )
-
-              val cons = consChain(elemsCells)
-              rewriter.solverContext.assertGroundExpr(tla.apalacheAssignChain(newSetCell.toNameEx, cons))
+          def addOneElemCons(elemCell: ArenaCell): Unit = {
+            def isPointedBySet(set: ArenaCell, setElems: Set[ArenaCell]): Boolean = setElems.contains(elemCell)
```

Comment:
Not sure what this will do. `setElems` is already a `Set[ArenaCell]`.

I imagine you meant to have `val setElemsAsSet = elemsOfSets.toSet` outside `addOneElemCons`, and use `setElemsAsSet` instead of `elemsOfSets` in the definition of `pointingSets`. Doesn't the call to `isPointedBySet` typecast to `Set[ArenaCell]` regardless?

Reply:
Oh, `setElems` is already a set. I got confused. Ignore the above comments then!